src/Codegen/Codegen.hs

```diff
@@ -651,7 +651,9 @@ genExpr (I.Match s as t) = do
       destruct <- getsDCon dconDestruct dcon
       cas      <- getsDCon dconCase dcon
       let fieldBinds =
-            zipWith (\field i -> (field, destruct i scrut)) fields [0 ..]
+            zipWith (\field i -> (getAltDefault field, destruct i scrut)) fields [0 ..]
+            where getAltDefault (I.AltDefault b) = b
+                  getAltDefault _ = Nothing
```

Comment:
Things like this should not be returning `Nothing`. In fact, they should be a compiler error/bug.

@j-hui I think we can just call the `error` function here?

Reply:
We can maybe create a function called `assumeAltDefault` in `IR.hs`, and use that function in the IR passes that assumes non-recursive patterns. This way encapsulates the `error` logic in that one place.

Reply:
Yeah, agree with @LeoQiao18 here